clients/src/main/java/org/apache/kafka/common/network/SslTransportLayer.java

```diff
@@ -903,6 +901,54 @@ private void updateBytesBuffered(boolean madeProgress) {
 
     @Override
     public long transferFrom(FileChannel fileChannel, long position, long count) throws IOException {
-        return fileChannel.transferTo(position, count, this);
+        if (state == State.CLOSING)
+            throw closingException();
+        if (state != State.READY)
+            return 0;
+
+        if (!flush(netWriteBuffer))
+            return 0;
+
+        long channelSize = fileChannel.size();
+        if (position > channelSize)
+            return 0;
+        int bytesToWrite = (int) Math.min(Math.min(count, channelSize - position), Integer.MAX_VALUE);
+
+        if (fileChannelBuffer == null) {
+            // same size as `netWriteBuffer`
+            int transferSize = 16384;
+            // Allocate a direct buffer to avoid one heap to heap buffer copy. SSLEngine copies the source
+            // buffer (fileChannelBuffer) to the destination buffer (netWriteBuffer) and then encrypts in-place.
+            // FileChannel.read() to a heap buffer requires a copy from a direct buffer to a heap buffer, which is not
+            // useful here.
+            fileChannelBuffer = ByteBuffer.allocateDirect(transferSize);
```

Comment:
If you allocate this as a direct buffer here, you need to force it to be deallocated in `SslTransportLayer#close`.  Otherwise these off-heap buffers will build up over time and starve the system of memory.  The garbage collector doesn't "see" direct buffers and in at least a few versions of Java, they never get cleaned up until a full GC.

Reply:
Good point.